mm2src/mm2_libp2p/src/request_response.rs

```diff
@@ -0,0 +1,262 @@
+            RequestResponseEvent::OutboundFailure {
+                peer,
+                request_id,
+                error,
+            } => {
+                error!("Error on send request {:?} to peer {:?}: {:?}", request_id, peer, error);
+                let err_response = PeerResponse::Err {
+                    err: format!("{:?}", error),
+                };
+                self.process_response(request_id, err_response);
+                return;
+            },
+        };
+
+        match message {
+            RequestResponseMessage::Request { request, channel } => {
+                debug!("Received a request from {:?} peer", peer_id);
+                self.process_request(peer_id, request, channel)
+            },
+            RequestResponseMessage::Response { request_id, response } => {
+                debug!(
+                    "Received a response to the {:?} request from peer {:?}",
+                    request_id, peer_id
+                );
+                self.process_response(request_id, response)
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Codec;
+
+#[derive(Debug, Clone)]
+pub enum Protocol {
+    Version1,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PeerRequest {
+    pub req: Vec<u8>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+pub enum PeerResponse {
+    Ok { res: Vec<u8> },
+    None,
+    Err { err: String },
+}
+
+macro_rules! try_io {
+    ($e: expr) => {
+        match $e {
+            Ok(ok) => ok,
+            Err(err) => return Err(io::Error::new(io::ErrorKind::InvalidData, err)),
+        }
+    };
+}
+
+impl ProtocolName for Protocol {
+    fn protocol_name(&self) -> &[u8] {
+        match self {
+            Protocol::Version1 => b"/request-response/1",
+        }
+    }
+}
+
+#[async_trait]
+impl RequestResponseCodec for Codec {
+    type Protocol = Protocol;
+    type Request = PeerRequest;
+    type Response = PeerResponse;
+
+    async fn read_request<T>(&mut self, _protocol: &Self::Protocol, io: &mut T) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        read_to_end(io).await
+    }
+
+    async fn read_response<T>(&mut self, _protocol: &Self::Protocol, io: &mut T) -> io::Result<Self::Response>
+    where
+        T: AsyncRead + Unpin + Send,
+    {
+        read_to_end(io).await
+    }
+
+    async fn write_request<T>(&mut self, _protocol: &Self::Protocol, io: &mut T, req: Self::Request) -> io::Result<()>
+    where
+        T: AsyncWrite + Unpin + Send,
+    {
+        write_all(io, &req).await
+    }
+
+    async fn write_response<T>(&mut self, _protocol: &Self::Protocol, io: &mut T, res: Self::Response) -> io::Result<()>
+    where
+        T: AsyncWrite + Unpin + Send,
+    {
+        write_all(io, &res).await
+    }
+}
+
+async fn read_to_end<T, M>(io: &mut T) -> io::Result<M>
+where
+    T: AsyncRead + Unpin + Send,
+    M: DeserializeOwned,
+{
+    match read_one(io, 1024).await {
```

Comment:
Will have to change the 1024 bytes limit to something more reasonable for example to request orderbook. As discussed at Discord we might want to fetch 20-30 best bids and asks from relayers so the size of requested data will exceed 1024 bytes.